.github/workflows/main.yml

```diff
@@ -1,61 +1,100 @@
+  # 
+  #   build:
+  #     runs-on: ubuntu-latest
+  #     strategy:
+  #       matrix:
+  #         os: [fedora, debian-x86_64, arch, debian-i386]
+  #     steps:
+  #       - uses: actions/checkout@v2
+  #       - name: Docker login
+  #         run: docker login docker.pkg.github.com -u $GITHUB_ACTOR -p $GITHUB_TOKEN
+  #         env:
+  #           GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+  #       - name: Build in container
+  #         env:
+  #           CI_NETWORK: true
+  #           CI: true
+  #         run: |
+  #           echo $GITHUB_WORKSPACE
+  #           docker run --privileged -e CI=true -t -v $GITHUB_WORKSPACE:/github/workspace docker.pkg.github.com/fwupd/fwupd/fwupd-${{matrix.os}}:latest
+  # 
+  #   fuzzing:
+  #     runs-on: ubuntu-latest
+  #     steps:
+  #     - name: Build Fuzzers
+  #       id: build
+  #       uses: google/oss-fuzz/infra/cifuzz/actions/build_fuzzers@master
+  #       with:
+  #         oss-fuzz-project-name: 'fwupd'
+  #         dry-run: false
+  #     - name: Run Fuzzers
+  #       uses: google/oss-fuzz/infra/cifuzz/actions/run_fuzzers@master
+  #       with:
+  #         oss-fuzz-project-name: 'fwupd'
+  #         fuzz-seconds: 300
+  #         dry-run: false
+  #     - name: Upload Crash
+  #       uses: actions/upload-artifact@v1
+  #       if: failure() && steps.build.outcome == 'success'
+  #       with:
+  #         name: artifacts
+  #         path: ./out/artifacts
 
-  build:
-    runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        os: [fedora, debian-x86_64, arch, debian-i386]
+  build-freebsd:
+    runs-on: macos-latest
+    name: Build fwupd from FreeBSD ports
     steps:
-      - uses: actions/checkout@v2
-      - name: Docker login
-        run: docker login docker.pkg.github.com -u $GITHUB_ACTOR -p $GITHUB_TOKEN
-        env:
-          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
-      - name: Build in container
-        env:
-          CI_NETWORK: true
-          CI: true
+    - name: Build
+      id: test
+      uses: vmactions/freebsd-vm@v0.1.2
+      with:
+        usesh: true
+        mem: 8192
+        prepare: |
+          pkg install -y git python3 glib meson pkgconf gobject-introspection \
+            vala gtk-doc json-glib gpgme gnutls sqlite3 curl gcab libarchive
+        sync: rsync
         run: |
-          echo $GITHUB_WORKSPACE
-          docker run --privileged -e CI=true -t -v $GITHUB_WORKSPACE:/github/workspace docker.pkg.github.com/fwupd/fwupd/fwupd-${{matrix.os}}:latest
+          mkdir -p /usr/local/etc/pkg/repos/
+          cp /etc/pkg/FreeBSD.conf /usr/local/etc/pkg/repos/FreeBSD.conf
+          sed -i .old 's|url: "pkg+http://pkg.FreeBSD.org/${ABI}/quarterly"|url: "pkg+http://pkg.FreeBSD.org/${ABI}/latest"|' /usr/local/etc/pkg/repos/FreeBSD.conf
+          pkg install -y libtasn1
+          pkg upgrade -y libtasn1
+          cd /usr
+          git clone https://github.com/3mdeb/freebsd-ports.git --depth 1 -b fwupd-port ports
```

Comment:
A few comments I notice about this:
1. It seems like `sysutils/fwupd/files/patch-libfwupd_meson.build` should be submitted to apply upstream.  The other ones too for `#ifdef` should each be reviewed too.
2. `DISTVERSION` I doubt you want to hardcode.  For the purpose of CI you should probably be passing this in as an environment variable or with `sed` or something. 
3. Overall I think it would be more flexible if we could keep an "upstream" copy of this BSD Makefile in the fwupd tree and you regularly sync it with the downstream tree.  Otherwise we may end up with a situation that we break the BSD build by hard bumping some dep but not being able to get the downstream BSD Makefile updated.  If we can keep it (or a template of it) upstream then we can just submit the changes to make sure CI keeps working for BSD..

Reply:
1. @Asiderr was the patch finally merged into fwupd/fwupd? By my understanding, the patches were for development purposes and now should be upstreamed.
2. I will update that to use current tag as DISTVERSION
3. My idea was that the CI would use the official freebsd-ports set to specific release (eg. 2020Q4) which would be updated manually. @Asiderr is working on upstreaming fwupd Makefile to freebsd/freebsd-ports and AFAIK it should be done soon.